### PR TITLE
Add support for Slurm states and dynamic hosts

### PR DESCRIPTION
**Pull request by jrha**
Dynamic host support requested by @apdibbo.
Slurm support requested by @drossy.

### CODE REVIEW

File: views/view-logical-workers.php [hunk 1]
```diff
@@ -68,9 +68,26 @@
...
 
+// Render all remaining hosts from the batch state table
+// this catches any hosts from cloud-bursting or other dynamic sources
+foreach ($all_status as $hostname => $status) {
+    if ($hostname !== '') {
+        $source = '';
+        if (sizeof($status) > 0) {
+            $source = array_values($status);
+            $source = $source[0];
+        }
```
Comment (tofu-rocketry) [inline views/view-logical-workers.php:84]: No semicolon needed? Or are the ones below the unneeded ones?

File: node/node-batchsystem.inc.php [hunk 2]
```diff
@@ -30,60 +30,32 @@ function detail($node, $short, $jobId = null, $jobCount = null)
...
         }
-        else {
-          echo "        <dt>Jobs</dt><dd>Not implemented</dd>\n";
+        for($i = 0; $i < $job_number; $i++) {
+          $state_visual = $state_visual . '<div class="cpu '.$dead.'inuse"></div>';
         }
+        for($i = 0; $i < ($slots - $job_number); $i++) {
+          $state_visual = $state_visual . '<div class="cpu '.$dead.'free"></div>';
+        }
+        echo "         <dt>Visual State</dt><dd>$state_visual</dd>\n";
```
Comment (tofu-rocketry) [inline node/node-batchsystem.inc.php:56]: What's with the funky spacing?

File: views/view-logical-workers.php (hunk 1 above)
Comment (jrha) [reply views/view-logical-workers.php:84]: Not needed… PHP…

File: node/node-batchsystem.inc.php (hunk 2 above)
Comment (jrha) [reply node/node-batchsystem.inc.php:56]: I think it was to aid debugging when reading the generated HTML.

File: views/view-logical-workers.php (hunk 1 above)
Comment (tofu-rocketry) [reply views/view-logical-workers.php:84]: What are the lower ones needed for then? :confused:

Comment (jrha) [reply views/view-logical-workers.php:84]: To keep the code nice and crufty?

Comment (tofu-rocketry) [reply views/view-logical-workers.php:84]: :man_facepalming:

**Review by tofu-rocketry — APPROVED** [review]
Probably fine? I mean, it's PHP, so probably not, but we'll ignore that for the moment.

File: views/view-logical-workers.php (hunk 1 above)
Comment (jrha) [reply views/view-logical-workers.php:84]: I've removed them.

File: node/node-batchsystem.inc.php (hunk 2 above)
Comment (tofu-rocketry) [reply node/node-batchsystem.inc.php:56]: Ok.

**Pull request merged**